File: src/android/LaunchNavigator.java
package org.apache.cordova.launchnavigator;

import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CallbackContext;

import org.json.JSONArray;
import org.json.JSONException;

import android.content.Intent;
import android.net.Uri;
import android.util.Log;

public class LaunchNavigator extends CordovaPlugin {

    private static final String TAG = LaunchNavigator.class.getName();
    private static final String TYPE_BAIDU = "baidu";
    private static final String TYPE_AMAP = "amap";


    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        boolean result;

        Log.d(TAG, action);

        if ("doLaunch".equals(action)) {
            result = doLaunch(args, callbackContext);
        } else {
            Log.e(TAG, "Invalid action");
            result = false;
        }

        if (result) {
            callbackContext.success();
        }
        return result;
    }

    private boolean doLaunch(JSONArray args, CallbackContext callbackContext) {
        boolean result = false;
        try {

            String type = args.getString(0);

            JSONArray dest = args.optJSONArray(1);
            JSONArray orig = args.optJSONArray(2);

            if (null != dest && null != orig) {
                result = this.doNavigate(type, dest.getString(0), dest.getString(1), dest.getString(2)
                        , orig.getString(0), orig.getString(1), orig.getString(2), callbackContext);
            } else if (null != dest) {
                result = this.doViewMap(type, dest.getString(0), dest.getString(1), dest.getString(2), callbackContext);
            }

        } catch (JSONException e) {
            Log.e(TAG, "Exception occurred: ".concat(e.getMessage()));
        }

        return result;
    }


    private boolean doViewMap(String type, String name, String lng, String lat, CallbackContext callbackContext) {
        boolean result;
        try {
            if (type.equals(TYPE_BAIDU)) {
                String url = String.format("bdapp://map/marker?location=%s,%s&title=%s&content=%s&coord_type=gcj02&src=duduche|parking"
                        , lat, lng, name, name);
                Log.d(TAG, url);
                Uri uri = Uri.parse(url);
                Intent intent = new Intent("android.intent.action.VIEW", uri);
                intent.setPackage("com.baidu.BaiduMap");
                cordova.getActivity().startActivity(intent);
                result = true;
            } else if (type.equals(TYPE_AMAP)) {
                String url = String.format("androidamap://viewMap?sourceApplication=duduche&lat=%s&lon=%s&poiname=%s&dev=0"
                        , lat, lng, name);
                Log.d(TAG, url);
                Uri uri = Uri.parse(url);
                Intent intent = new Intent("android.intent.action.VIEW", uri);
                intent.setPackage("com.autonavi.minimap");
                cordova.getActivity().startActivity(intent);

                result = true;
            } else {
                callbackContext.error("Invalid type.");
                result = false;
            }
        } catch (Exception e) {
            String msg = e.getMessage();
            Log.e(TAG, "Exception occurred: ".concat(msg));
            callbackContext.error(msg);
            result = false;
        }

        return result;
    }


    private boolean doNavigate(String type, String dname, String dlat, String dlng
            , String oname, String olat, String olng, CallbackContext callbackContext) {
        boolean result;
        try {
            if (type.equals("baidu")) {
                String url = String.format("bdapp://map/direction?origin=%s,%s&destination=%s,%s&coord_type=gcj02&mode=driving&src=duduche|parking"
                        , olat, olng, dlat, dlng);
                Log.d(TAG, url);
                Uri uri = Uri.parse(url);
                Intent intent = new Intent("android.intent.action.VIEW", uri);
                intent.setPackage("com.baidu.BaiduMap");
                cordova.getActivity().startActivity(intent);
                result = true;
            } else if (type.equals("amap")) {
                String url = String.format("androidamap://route?sourceApplication=duduche&slat=%s&slon=%s&sname=%s&dlat=%s&dlon=%s&dname=%s&dev=0&m=0&t=2"
                        , olat, olng, oname, dlat, dlng, dname);
                Log.d(TAG, url);
                Uri uri = Uri.parse(url);
                Intent intent = new Intent("android.intent.action.VIEW", uri);
                intent.setPackage("com.autonavi.minimap");
                cordova.getActivity().startActivity(intent);

                result = true;
            } else {
                callbackContext.error("Invalid type.");
                result = false;
            }
        } catch (Exception e) {
            String msg = e.getMessage();
            Log.e(TAG, "Exception occurred: ".concat(msg));
            callbackContext.error(msg);
            result = false;
        }

        return result;
    }

}

File: src/ios/LaunchNavigator.m
//
//  lalaPlugin.m
//  hello
//
//  Created by seasu on 2015/7/29.
//
//

#import "LaunchNavigator.h"
#import <Cordova/CDV.h>


@implementation LaunchNavigator

NSString *TYPE_BAIDU = @"baidu";
NSString *TYPE_AMAP = @"amap";


- (void)doLaunch:(CDVInvokedUrlCommand*)command {
    
    CDVPluginResult *pluginResult = nil;
    
    NSString *type = [command argumentAtIndex:0];
    
    NSArray *dest = [command argumentAtIndex:1];
    NSArray *orig = [command argumentAtIndex:2];
    
    if (type == nil) {
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"navigator type missing"];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        return;
    }
    if ([@[TYPE_AMAP, TYPE_BAIDU] containsObject:type] == NO) {
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"navigator type not support1"];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        return;
    }
    
    @try {
        if (dest != nil && orig != nil) {
            [self navigate:type originName:orig[0] originLat:orig[2] originLng:orig[1] destinationName:dest[0] destinationLat:dest[2] destinationLng:dest[1] command:command];
        } else if (dest != nil){
            [self viewMap:type title:dest[0] lat:dest[2] lng:dest[1] command:command];
        } else {
            pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"navigator dest missing"];
            [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
            return;
        }
        }
    @catch (NSException *exception) {
    }
    @finally {
    }
    
}

- (void)viewMap:(NSString *)type
          title:(NSString *)name
            lat:(NSString *)lat
            lng:(NSString *)lng
        command:(CDVInvokedUrlCommand *)command {
    NSString *str = nil;
    if ([type isEqualToString:TYPE_BAIDU]) {
        str = [NSString stringWithFormat:@"baidumap://map/marker?location=%@,%@&title=%@&coord_type=gcj02&src=duduche|parking&content=目的地", lat, lng, name];
    } else if ([type isEqualToString:TYPE_AMAP]) {
        str = [NSString stringWithFormat:@"iosamap://viewMap?sourceApplication=duduche&poiname=%@&lat=%@&lon=%@&dev=0", name, lat, lng];
    }
    str = [str stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
    [self openUrl:[NSURL URLWithString:str] command:command];
}

- (void)navigate:(NSString *)type
      originName:(NSString *)originName
       originLat:(NSNumber *)originLat
       originLng:(NSNumber *)originLng
 destinationName:(NSString *)destinationName
  destinationLat:(NSNumber *)destinationLat
  destinationLng:(NSNumber *)destinationLng
         command:(CDVInvokedUrlCommand *)command
{
    NSString *str = nil;
    if ([type isEqualToString:TYPE_BAIDU]) {
        str = [NSString stringWithFormat:@"baidumap://map/direction?origin=%@,%@&destination=%@,%@&mode=driving&src=duduche", originLat, originLng, destinationLat, destinationLng];
    } else if ([type isEqualToString:TYPE_AMAP]) {
        str = [NSString stringWithFormat:@"iosamap://path?sourceApplication=duduche&sname=%@&slat=%@&slon=%@dname=%@&dlat=%@&dlon=%@&dev=0&m=0&t=0", originName, originLat, originLng, destinationName, destinationLat, destinationLng];
    }
    str = [str stringByAddingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
    [self openUrl:[NSURL URLWithString:str] command:command];
}

- (void)openUrl:(NSURL *)url command:(CDVInvokedUrlCommand *)command{
    CDVPluginResult *pluginResult = nil;
    NSLog(@"%@", url.absoluteString);
    if ([[UIApplication sharedApplication] canOpenURL:url]) {
        [[UIApplication sharedApplication] openURL:url];
    } else {
        pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"ios application can not open."];
        [self.commandDelegate sendPluginResult:pluginResult callbackId:command.callbackId];
        return;
    }
}

@end
